stability: expose (and fix) data races in Transport

tbg opened the pull request:
A BatchRequest may be passed on to multiple goroutines within the kv.Transport,
and so effectively `DistSender` and `TxnCoordSender` should not assume that
they have exclusive access to any of the pointer fields in it. This was not
always obeyed and various data races are fixed in this commit.

They were identified with a newly introduced mechanism, which during
race-enabled builds intercepts all BatchRequests passing through GRPCTransport
and continuously accesses them. This vastly improving race detector coverage in
our test suite for this type of race.

Most of the races are directly related to updating a `Transaction` without
copy-on-write; one highlighted that we should copy the request in the transport
even when bypassing GRPC completely (for requests which go to the local Node).

At times, it was necessary to increase the race detector's history size to
capture the offending goroutines' stack traces:

```
make testrace GORACE=history_size=7 PKG=./sql TESTS=TestAsOfRetry
```

Fixes #8020.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8865)

<!-- Reviewable:end -->

tamird commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[kv/transport.go, line 287 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUVsU7PGAl_qSmnvq:-KQAUVsU7PGAl_qSmnvr:b-iawnlv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L287)):*
> ```Go
> 		s SendOptions, c *rpc.Context, sl ReplicaSlice, args roachpb.BatchRequest,
> 	) (Transport, error) {
> 		_ = stopper.RunAsyncTask(func() {
> ```

for fun: check this error. it can lead to surprises

---

*[kv/transport.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUhppAhFpP26A0jIx:-KQAUhpq5L8QveJK31MI:bpmsksi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L298)):*
> ```Go
> 					// all fields (it would be enough to access all pointer
> 					// fields as the rest has been passed by-value anyway).
> 					if err := encoder.Encode(&args); err != nil {
> ```

Have you tried doing something cheaper? For example, `reflect.DeepEqual` against some randomly generated transaction struct? That might alleviate some other comments about this being too slow.

---

*[kv/transport.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWX_lJyre8Uq9kCRq:-KQAWX_lJyre8Uq9kCRr:b-irtpiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L327)):*
> ```Go
> 
> func (s *racePromotingTransport) SendNext(done chan BatchCall) {
> 	tp := s.factory()
> ```

something about this doesn't seem right. why is it ok to create a new wrapped transport on every call to SendNext while also ignoring the wrapped transport's `IsExhausted`?

---

*[server/server.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUCn1yjasecxZXR5v:-KQAUCn2THyMN7Wkfs2X:bshj0iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L77)):*
> ```Go
> // system.
> //
> // It is not called ServerTestingKnobs to satisfy the linters.
> ```

which linter?

---

*[server/server.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWAh95VU-XEEYpddu:-KQAWAh95VU-XEEYpddv:bylqyw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L194)):*
> ```Go
> 		// make things an awful lot slower; perhaps the reads don't need to
> 		// be as aggressively timed.
> 		_ = kv.RacePromotingTransportFactory(kv.GRPCTransportFactory, s.stopper)
> ```

why special case it this way rather than put it in testingknobs?

---

*[server/server.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAVqjSW4-E1HbK3k9C:-KQAVqjSW4-E1HbK3k9D:b5enw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L196)):*
> ```Go
> 		_ = kv.RacePromotingTransportFactory(kv.GRPCTransportFactory, s.stopper)
> 	}
> 	if serverKnobs, _ := srvCtx.TestingKnobs.Server.(*SrvTestingKnobs); serverKnobs != nil {
> ```

why not use `ok`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[kv/transport.go, line 287 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUVsU7PGAl_qSmnvq:-KQBTMu6p92TogS58syc:b-q296f4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> for fun: check this error. it can lead to surprises
</details>
What would I do with it? I certainly expect to see the error every now and then during shutdown.

---

*[kv/transport.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUhppAhFpP26A0jIx:-KQBTWIazyrtnEgRwUgI:binn4ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L298)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Have you tried doing something cheaper? For example, `reflect.DeepEqual` against some randomly generated transaction struct? That might alleviate some other comments about this being too slow.
</details>
`reflect.DeepEqual` would quit very early on (since it's... not equal). I could speculate that `reflect.DeepEqual(ba, ba)` would do 100% of the work, though, but  I'd rather stick with something that's more obviously correct.

I'll have to play with speeding it up. I think the transport could be made more efficient if it had a single worker (i.e. the more pending requests, the less iterations per request). Will take a look at that option.

---

*[kv/transport.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWX_lJyre8Uq9kCRq:-KQBTt2Ej_jw-pHz-Rl4:b2lv97k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L327)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> something about this doesn't seem right. why is it ok to create a new wrapped transport on every call to SendNext while also ignoring the wrapped transport's `IsExhausted`?
</details>
You're right that this code is slightly incorrect to be simpler. I fixed it up.

---

*[server/server.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUCn1yjasecxZXR5v:-KQBUhtEdQciKGdJllOs:b1sobfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L77)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> which linter?
</details>
The stuttering warning. I believe that's `golint`?

---

*[server/server.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWAh95VU-XEEYpddu:-KQBUt_U26ZSfiU2ilth:b4urnbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L194)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why special case it this way rather than put it in testingknobs?
</details>
What do you mean? The "real" test uses TestingKnobs but this is intermediate code which allows running *all* users of `TestServers` with that transport (a knob doesn't allow me that).

---

*[server/server.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAVqjSW4-E1HbK3k9C:-KQBV1xu5Z3zy6eNSJZn:b-st4ima) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L196)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not use `ok`?
</details>
Why use it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[kv/transport.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUhppAhFpP26A0jIx:-KQBaZxkbwL2gGuqRPBJ:bye5kwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L298)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `reflect.DeepEqual` would quit very early on (since it's... not equal). I could speculate that `reflect.DeepEqual(ba, ba)` would do 100% of the work, though, but  I'd rather stick with something that's more obviously correct.
> 
> I'll have to play with speeding it up. I think the transport could be made more efficient if it had a single worker (i.e. the more pending requests, the less iterations per request). Will take a look at that option.
</details>
(btw, for most individual tests the slowdown isn't that bad, so this is really only an issue if we want to enable this for all tests in race mode)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[kv/transport.go, line 287 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUVsU7PGAl_qSmnvq:-KQBjqaVhs4DaBxu-HNT:b-5gppl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L287)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What would I do with it? I certainly expect to see the error every now and then during shutdown.
</details>
Yeah, it's certainly expected, but we've had hard-to-track-down shutdown issues before, so it's usually good to do something with these errors. How about returning it from this method?

---

*[kv/transport.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUhppAhFpP26A0jIx:-KQBk3T2LlQq66G6Jb82:b-h7nwcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L298)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> (btw, for most individual tests the slowdown isn't that bad, so this is really only an issue if we want to enable this for all tests in race mode)
</details>
Seems like that might be worthwhile, no?

---

*[kv/transport.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWX_lJyre8Uq9kCRq:-KQBk8Jzztw8iv9H_Rdy:b-tcok12) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L327)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You're right that this code is slightly incorrect to be simpler. I fixed it up.
</details>
Publish before push?

---

*[server/server.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUCn1yjasecxZXR5v:-KQBkAL7oM0i3PamHVm5:b-4e4zi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L77)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The stuttering warning. I believe that's `golint`?
</details>
Ah, yeah, it's golint. Can you put that in the comment? s/satisfy the linters/avoid stuttering/

---

*[server/server.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWAh95VU-XEEYpddu:-KQBkJMTcHt3MGz-PzID:b-wwc9pw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L194)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What do you mean? The "real" test uses TestingKnobs but this is intermediate code which allows running *all* users of `TestServers` with that transport (a knob doesn't allow me that).
</details>
Perhaps this becomes clearer after your update.

---

*[server/server.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAVqjSW4-E1HbK3k9C:-KQBkVVQhs4n5eeD2bU_:b38d0z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L196)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why use it?
</details>
Because explicit is better than implicit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 6 of 8 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[kv/transport.go, line 287 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUVsU7PGAl_qSmnvq:-KQBoGKL-nhVL3uT_03P:bm4pggi) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L287)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yeah, it's certainly expected, but we've had hard-to-track-down shutdown issues before, so it's usually good to do something with these errors. How about returning it from this method?
</details>
That's easy enough, done.

---

*[kv/transport.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUhppAhFpP26A0jIx:-KQBp8t6ilJatlbAFwUd:b-vyjf3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L298)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Seems like that might be worthwhile, no?
</details>
Yes, I think so. In particular the `sql` test suite should give this code a good scrub.

---

*[kv/transport.go, line 327 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWX_lJyre8Uq9kCRq:-KQBoKkoMPqiC0DesyT-:bo6apac) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L327)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Publish before push?
</details>
Indeed.

---

*[server/server.go, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUCn1yjasecxZXR5v:-KQBoREy2DGxhbrrRqUH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L77)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Ah, yeah, it's golint. Can you put that in the comment? s/satisfy the linters/avoid stuttering/
</details>
Done.

---

*[server/server.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAVqjSW4-E1HbK3k9C:-KQBoUh6DxUqyAZYx2jk:b-h6pshl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L196)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Because explicit is better than implicit.
</details>
That (candidate?) Go proverb certainly has merit, but in this situation I think we're really talking about personal preference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[kv/transport.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUhppAhFpP26A0jIx:-KQBsbZ84BtTLSfirHSZ:b-w1r1jp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L298)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yes, I think so. In particular the `sql` test suite should give this code a good scrub.
</details>
Meaning you want to do it in a subsequent change? FWIW, `reflect.DeepEqual(args,args)` may also not do what you want if it's smart enough to bail after the first pointer comparison.

---

*[kv/transport.go, line 342 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQBrWN0h66nPAULzExT:-KQBrWN0h66nPAULzExU:b-st8skk) ([raw file](https://github.com/cockroachdb/cockroach/blob/448f0c4b2fefc742abb910dabb9cb73dd86c95e2/kv/transport.go#L342)):*
> ```Go
> 	}
> 	s.curTP.SendNext(done)
> 	s.curTP = nil
> ```

why does this get cleared and recreated on each `IsExhausted/SendNext` cycle?

---

*[server/server.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWAh95VU-XEEYpddu:-KQBs5Sr_YrVlfwyuOOS:b-68xc62) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L194)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Perhaps this becomes clearer after your update.
</details>
OK, so I'm still missing something here. How can you tell that this only happens if TestServer is in us? Seems like you're doing this no matter what, and in that case you may as well just `tranportFactory := kv.RacePromotingTransportFactory(kv.GRPCTransportFactory, s.stopper)` unconditionally.

That probably won't work right - see my other comment about the wrapped transport - but that should be fixed anyway.

---

*[server/server.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAVqjSW4-E1HbK3k9C:-KQBrkta20yyGddMAuT1:bwyb76q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L196)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> That (candidate?) Go proverb certainly has merit, but in this situation I think we're really talking about personal preference.
</details>
Fine, but consider that the pattern recognition part of my brain has to work a little harder than if this used `ok`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 3 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[kv/transport.go, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAUhppAhFpP26A0jIx:-KQBttbqMjXVTMT2Fzta:bqjicay) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/kv/transport.go#L298)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Meaning you want to do it in a subsequent change? FWIW, `reflect.DeepEqual(args,args)` may also not do what you want if it's smart enough to bail after the first pointer comparison.
</details>
No, doing it here. Yes, that `DeepEqual` was going to try to be smart was also my concern. With the latest changes (single-goroutine read loop) it seems that performance doesn't suffer that much.

I pushed a new commit which makes GRPCTransport race-enabled, so all of the manual stuff is now gone. It caught a few new races, but I'm going to push anyway (so CI will be red).

---

*[kv/transport.go, line 342 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQBrWN0h66nPAULzExT:-KQBuEa9qJZvfd2DRvYK:b4tnhcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/448f0c4b2fefc742abb910dabb9cb73dd86c95e2/kv/transport.go#L342)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why does this get cleared and recreated on each `IsExhausted/SendNext` cycle?
</details>
Because we didn't at all want the transport to be exhausted (and using an exhausted transport is illegal). But with the latest change I don't think I need that. PTAL, this simplified it dramatically.

---

*[server/server.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAWAh95VU-XEEYpddu:-KQBt0xBIFhZF7wYyb4j:b-5jrboc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L194)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK, so I'm still missing something here. How can you tell that this only happens if TestServer is in us? Seems like you're doing this no matter what, and in that case you may as well just `tranportFactory := kv.RacePromotingTransportFactory(kv.GRPCTransportFactory, s.stopper)` unconditionally.
> 
> That probably won't work right - see my other comment about the wrapped transport - but that should be fixed anyway.
</details>
The line as written is a no-op as the transport isn't used. If you assign it to `transportFactory` (i.e. alter the line), it is always used (except when another serverKnobs is specified). And I confused myself (and maybe you) with that TestServer statement: s/TestServer/Server/ is what I wanted.

---

*[server/server.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQAVqjSW4-E1HbK3k9C:-KQBtVDonlVZO1IS5jqX:baggbl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca9c15abedac731be7cbe44e1b1b1dddbf760c1e/server/server.go#L196)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Fine, but consider that the pattern recognition part of my brain has to work a little harder than if this used `ok`.
</details>
And I have a pattern recognition that works the other way, or I wouldn't have written it like that. I assume what you want is

```
if x, ok := some.(*thing); ok {
}
```

and that triggers wanting to check `x != nil` and then we arrive at the unwieldy `ok && x != nil` which is too verbose.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
With another race fix I just pushed, this is now green.

tamird commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[kv/transport.go, line 117 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQDJy-eKWJfWBcN8WlW:-KQDJy-fyg8hNK8BUUjX:b23ucyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd02b488e94435f90eb7bbbd7cc90aa8ffdf6a8/kv/transport.go#L117)):*
> ```Go
> // During race builds, it does extra work to hold on to and read all obtained
> // requests in a tight loop, exposing data races.
> var GRPCTransportFactory = func() func(
> ```

can probably return this to being unexported. also, any reason for this to be a `var`?

---

*[kv/transport.go, line 123 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQDIj70z9LiZTn6uUyP:-KQDIj70z9LiZTn6uUyQ:b-hj9lov) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd02b488e94435f90eb7bbbd7cc90aa8ffdf6a8/kv/transport.go#L123)):*
> ```Go
> 	args roachpb.BatchRequest,
> ) (Transport, error) {
> ```

spurious line

---

*[kv/transport.go, line 194 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQDIuIVZhLi7xAaNH3m:-KQDIuIVZhLi7xAaNH3n:bowz0he) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd02b488e94435f90eb7bbbd7cc90aa8ffdf6a8/kv/transport.go#L194)):*
> ```Go
> 		//
> 		// TODO(tamird): we should clone all of client.args.Header, but the
> 		// assertions in protoutil.Clone fire and there seems to be no
> ```

the workaround is to implement Clone yourself :(

---

*[server/server.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQDJe8XEOii2jH_8myV:-KQDJe8XEOii2jH_8myW:b-pvl2jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd02b488e94435f90eb7bbbd7cc90aa8ffdf6a8/server/server.go#L174)):*
> ```Go
> 	retryOpts := base.DefaultRetryOptions()
> 	retryOpts.Closer = s.stopper.ShouldQuiesce()
> ```

spurious whitespace-only change in this file now

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-:-KQDJmFHc5Rr7a0BiyKH:b-uwkowd)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 2 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[kv/transport.go, line 117 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQDJy-eKWJfWBcN8WlW:-KQEB_d91r_KgyQfWwro:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd02b488e94435f90eb7bbbd7cc90aa8ffdf6a8/kv/transport.go#L117)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can probably return this to being unexported. also, any reason for this to be a `var`?
</details>
Done.

---

*[kv/transport.go, line 123 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQDIj70z9LiZTn6uUyP:-KQEB_FgtNiNNZMIGMew:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd02b488e94435f90eb7bbbd7cc90aa8ffdf6a8/kv/transport.go#L123)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> spurious line
</details>
Done.

---

*[server/server.go, line 174 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQDJe8XEOii2jH_8myV:-KQE5Mbm5K99pBFfJ1q6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5fd02b488e94435f90eb7bbbd7cc90aa8ffdf6a8/server/server.go#L174)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> spurious whitespace-only change in this file now
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tbg committed:
stability: expose (and fix) data races in Transport

A BatchRequest may be passed on to multiple goroutines within the kv.Transport,
and so effectively `DistSender` and `TxnCoordSender` should not assume that
they have exclusive access to any of the pointer fields in it. This was not
always obeyed and various data races are fixed in this commit.

They were identified with a newly introduced mechanism, which during
race-enabled builds intercepts all BatchRequests passing through GRPCTransport
and continuously accesses them. This vastly improving race detector coverage in
our test suite for this type of race.

Most of the races are directly related to updating a `Transaction` without
copy-on-write; one highlighted that we should copy the request in the transport
even when bypassing GRPC completely (for requests which go to the local Node).

At times, it was necessary to increase the race detector's history size to
capture the offending goroutines' stack traces:

    make testrace GORACE=history_size=7 PKG=./sql TESTS=TestAsOfRetry

Fixes #8020.

tamird commented:
Reviewed 6 of 6 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[kv/transport_race.go, line 37 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFvrxBjZEeaHx-SnNU:-KQFvrxBjZEeaHx-SnNV:bphb9bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L37)):*
> ```Go
> 
> var running atomic.Value // bool
> var incoming chan *roachpb.BatchRequest
> ```

looking at this with fresher eyes, I think you can use `roachpb.BatchRequest` (i.e. a value rather than pointer) throughout, since that's how it is passed around.

---

*[kv/transport_race.go, line 82 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFwEl7RzFHoSRiK4C7:-KQFwEl7RzFHoSRiK4C8:bkvvfzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L82)):*
> ```Go
> 					select {
> 					case ba := <-incoming:
> 						bas[curIdx] = ba
> ```

instead of holding a limited slice and overwriting, could you hook the transport's Close() and remove requests that way?

it may be easier to spin a goroutine for each transport after all?

---

*[kv/transport_race.go, line 91 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFusBKEC5TraemVCMk:-KQFusBLjNF4HuM50z59:bc8aabz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L91)):*
> ```Go
> 				for _, ba := range bas {
> 					if ba != nil {
> 						if err := encoder.Encode(&ba); err != nil {
> ```

nit: this is a **roachpb.BatchRequest, which probably isn't necessary

---

*[kv/transport_race.go, line 99 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFvSlcZSCbkY3Ea0gA:-KQFvSlcZSCbkY3Ea0gB:bgpa5dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L99)):*
> ```Go
> 		})
> 	}
> 	select {
> ```

shouldn't this be a blocking send?

---

*[kv/txn_coord_sender.go, line 283 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFuRWTpv1jZVl8DDJ7:-KQFuRWTpv1jZVl8DDJ8:b-db7w3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/txn_coord_sender.go#L283)):*
> ```Go
> 			ba.Header.Trace = &tracing.Span{}
> 		} else {
> 			// This is unfortunately necessary to prevent a data race which is
> ```

can you elaborate on the mechanics? this comment is of the sort that becomes enshrined as immutable FUD.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r1, 4 of 6 files at r4, 1 of 1 files at r5, 2 of 2 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[kv/transport_race.go, line 60 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQI8JHh88phGz46XG8K:-KQI8JHh88phGz46XG8L:b3we2yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfda56953f4be57e10e1daa3d55c02326c352703/kv/transport_race.go#L60)):*
> ```Go
> ) (Transport, error) {
> 	if !running.Load().(bool) {
> 		running.Store(true)
> ```

This is not atomic and could allow multiple workers to start at once. This needs to be a single atomic operation instead of separate load and store. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-:-KQIA7j7tWKMU1eO3-QQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 6 of 8 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[kv/transport_race.go, line 37 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFvrxBjZEeaHx-SnNU:-KQH3Y6c7pbMW5Jf_eG9:b-aellvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L37)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> looking at this with fresher eyes, I think you can use `roachpb.BatchRequest` (i.e. a value rather than pointer) throughout, since that's how it is passed around.
</details>
I can, but I chose to use `*BatchRequest` on purpose as it's more idiomatic to what we're doing here (reading the same memory over and over). We know that some of the BatchRequest can't race (because we start out with a shallow copy), but I doubt it's worth optimizing further.

---

*[kv/transport_race.go, line 82 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFwEl7RzFHoSRiK4C7:-KQH43UzIZVj5CkUk2ng:b-v00zq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> instead of holding a limited slice and overwriting, could you hook the transport's Close() and remove requests that way?
> 
> it may be easier to spin a goroutine for each transport after all?
</details>
The transport's `Close()` is when the transport is destroyed. It is not, however, when we need to stop for data races. The offending write typically happens after the transport is destroyed.

---

*[kv/transport_race.go, line 91 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFusBKEC5TraemVCMk:-KQHQ04U8T1Zqyn4MAoS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L91)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: this is a **roachpb.BatchRequest, which probably isn't necessary
</details>
Done.

---

*[kv/transport_race.go, line 99 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFvSlcZSCbkY3Ea0gA:-KQH3m1bo60Mf5ITamF7:b-jnxedc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L99)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> shouldn't this be a blocking send?
</details>
I didn't want to be too invasive here. If we're already scrambling, no point in blocking the tests (which certainly costs a lot of CPU time).

---

*[kv/transport_race.go, line 60 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQI8JHh88phGz46XG8K:-KQIPSrOdx9YZQnYu2BH:bpu21ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/cfda56953f4be57e10e1daa3d55c02326c352703/kv/transport_race.go#L60)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is not atomic and could allow multiple workers to start at once. This needs to be a single atomic operation instead of separate load and store. 
</details>
It's not an issue if that happens, but I changed this to use an atomically updated int32 to avoid questions here.

---

*[kv/txn_coord_sender.go, line 283 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFuRWTpv1jZVl8DDJ7:-KQH3jx5xhpD1nKHdJEx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/txn_coord_sender.go#L283)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you elaborate on the mechanics? this comment is of the sort that becomes enshrined as immutable FUD.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@tamird, any last words?

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r10, 1 of 2 files at r12, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[kv/transport_race.go, line 37 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQFvrxBjZEeaHx-SnNU:-KQIkKVuas1XDfD3EBBK:bk0z7qo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6df7c5f84e8d61b9271932752b9ec79feaa7091f/kv/transport_race.go#L37)):*
> I can, but I chose to use *BatchRequest on purpose as it's more idiomatic to what we're doing here (reading the same memory over and over).

It looks more idiomatic at the cost of being misleading. I support I don't feel strongly enough to insist.

---

*[kv/transport_race.go, line 70 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-KQIkmdz7Qf3aFEjMooo:-KQIkmdz7Qf3aFEjMoop:b-c0xl5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b713e4362b78bd7068a9ffbb325738f79bf0cd/kv/transport_race.go#L70)):*
> ```Go
> 				)
> 			}()
> 			// Make a fixed-size slice of *BatchRequest which is populated
> ```

"populated FILO" doesn't make sense since "populated" seems only to refer to the "In" part. also, entries are evicted in FIFO, as far as I can tell.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-:-KQIlOpHOIe7mg0a9C5O:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
Bump default timeout for `make testrace` to 8m

The new race promotion in the GRPC transport pushes CircleCI over the edge.
In particular, TestLogic eats a lot of CPU time, though during local testing
it finishes in very much the same time as without these changes.

Since TeamCity also manages to get through `make testrace` in the previously
allotted five minutes, this is likely just a reflection of how little CPU is
available on CircleCI, and a higher timeout should help.

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 7 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8865#-:-KQLUvjsroVj3qrVCsY9:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
server: skip flaky TestAdminAPITableStats
